### PR TITLE
Ship yarn with hyper

### PR DESCRIPTION
**Pull request by sheerun**
This is a step towards #282 and makes sure hyperterm always
installs with the same npm version it shipped with.

It also drops dependency on system's node.

### CODE REVIEW

**Comment by rauchg** [comment]
That's really sweet. I wonder how much bigger it makes our compressed builds / startup time ?

**Comment by leo** [comment]
I'm not sure why you're ignoring [this](https://github.com/zeit/hyperterm/pull/172#issuecomment-233174361).

**Comment by sheerun** [comment]
It doesn't mater as you embed npm, and have can fix working version if you need.

Spawning a new process simply doesn't work because there's no node installed in electron app..

**Comment by rauchg** [comment]
I would love for @othiym23 to re-visit that decision. It's kind of a bummer that we need to ask the user to install node to use `npm` as an API :(

**Comment by sindresorhus** [comment]
> Spawning a new process simply doesn't work because there's no node installed in electron app..

You can. Just spawn `process.execPath` with the [`ELECTRON_RUN_AS_NODE`](https://github.com/electron/electron/blob/master/docs/api/environment-variables.md#electron_run_as_node) environment variable, and it will use the embedded Node.js.

```js
childProcess.execSync(process.execPath, ['npm', 'install'], {env: {ELECTRON_RUN_AS_NODE: true}});
```

**Comment by sheerun** [comment]
I tried to implement it as @sindresorhus proposed, but for some reason electron refuses to exit when running npm with it. Here's how to reproduce:

```bash
export CWD=$(pwd) && (cd ~/.hyperterm_plugins && ELECTRON_RUN_AS_NODE=1 $CWD/node_modules/electron-prebuilt/dist/Electron.app/Contents/MacOS/Electron $CWD/node_modules/.bin/npm install --verbose)
```

For now it's blocking me, so any help would be appreciated..

**Comment by rauchg** [comment]
@sheerun seems like something we should be reporting to the electron team. This is looking very good btw, thanks @sindresorhus

**Comment by rauchg** [comment]
@sheerun I'd love to merge this. Do you have any news about whether the issue with exiting was fixed? Or any workarounds?

**Comment by sheerun** [comment]
The issue to fix is here: https://github.com/electron/electron/issues/4944

With newest electron it kinda works, but takes 30 seconds to exit.

The best I can do is to rebase, catch.

**Comment by sheerun** [comment]
@leo @rauchg I've replaced npm with yarn and it runs like wonder. It's good to merge.

There's issue with CI build as `xo` doesn't recognize `queue` package as dependency (it is). Maybe that's something @sindresorhus could check..

**Comment by matheuss** [comment]
@sheerun `queue` is on the wrong `package.json` 😅 You need to move it to `app/package.json` 😄

**Comment by sheerun** [comment]
@matheuss Is `yarn` on correct one?

**Comment by sheerun** [comment]
Anyway, I fixed the issue with build. Ready for review and merge..

**Comment by sheerun** [comment]
@matheuss This is no longer in-progress. This is done.

**Comment by matheuss** [comment]
@sheerun thanks for the ping 😄 

> Is yarn on correct one?

Nope, you need to move it to the other one 👌 

Could you please rebase to fix the merge conflicts? I'll make sure to discuss it with the team 👌

**Comment by sheerun** [comment]
@matheuss @rauchg Done. I hope I've rebased it the last time... :)

**Comment by rauchg** [comment]
This is fantastic!

**Comment by sheerun** [comment]
there are conflicts again.... Can I ask you rebase and merge it by yourself?

**Comment by rauchg** [comment]
Yes. Sorry for the trouble

**Comment by timneutkens** [comment]
@sheerun After building this branch using `npm run pack` I'm seeing an issue related to
```
  const electronPath = resolve(
    __dirname, '..', 'node_modules', 'electron',
    readFileSync(resolve(__dirname, '..', 'node_modules', 'electron', 'path.txt')).toString().trim()
  );
```

path.txt can't be found. Can you give me some insights into this?

**Comment by sheerun** [comment]
@timneutkens electron-prebuit should install it in postinstall hook: https://github.com/electron-userland/electron-prebuilt/blob/master/install.js#L39

It uses this file even for executing itself, so this file must be present: https://github.com/electron-userland/electron-prebuilt/blob/master/index.js

**Comment by sheerun** [comment]
any progress?

**Comment by albinekb** [comment]
Yarn has a `--json` option:

kapture 2017-06-10 at 4 52 04

maybe that's a better way to run it?

**Comment by albinekb** [comment]
I think this is the better way to deal with this 👌  Update the deps (yarn 0.17 is old 🗡 ) and then we need to test it.

**Comment by sheerun** [comment]
With all respect, I've already reabased and updated this PR per request
like 5 times. I think this time it's up to you to update it.

**Comment by chabou** [comment]
Rebased and yarn version changed to 0.21.3: Since 0.22.0 it gives me errors:
```
App threw an error during load
TypeError: Cannot read property 'from' of undefined
    at Object.<anonymous> (/Users/chabou/Documents/Projets/hyper/node_modules/readable-stream/node_modules/safe-buffer/index.js:3:11)
    at Object.<anonymous> (/Users/chabou/Documents/Projets/hyper/node_modules/readable-stream/node_modules/safe-buffer/index.js:60:3)
    at Module._compile (/Users/chabou/Documents/Projets/hyper/node_modules/v8-compile-cache/v8-compile-cache.js:178:30)
    at Object.Module._extensions..js (module.js:580:10)
    at Module.load (module.js:488:32)
    at tryModuleLoad (module.js:447:12)
    at Function.Module._load (module.js:439:3)
    at Module.require (module.js:498:17)
    at require (/Users/chabou/Documents/Projets/hyper/node_modules/v8-compile-cache/v8-compile-cache.js:159:20)
    at Object.<anonymous> (/Users/chabou/Documents/Projets/hyper/node_modules/readable-stream/lib/_stream_readable.js:32:14)

```

**Comment by chabou** [comment]
Same error than @timneutkens when using built version.
There isn't neither `path.txt` file nor `node_modules` directory:
`Uncaught Exception:
Error: ENOENT: no such file or directory, open '/Users/chabou/Documents/Projets/hyper/dist/mac/Hyper.app/Contents/Resources/node_modules/electron/path.txt'`

**Comment by chabou** [comment]
I'm trying to debug this problem with `path.txt` file but there is an important thing to note. 
This PR can't work if we have `asar` activated: `node_modules` directory will be packed in  `app.asar` archive 😞

**Comment by chabou** [comment]
This PR do not work by design.
`path.txt` can't be used to this purpose. It almost [constant paths](https://github.com/electron-userland/electron-prebuilt/blob/master/install.js#L51-L63) that are only valid in dev mode. In built mode, we do not have electron dependency, we need to execute Hyper itself (as node_engine), not Electron.

I tried to use `asarUnpack` to extract only `./node_modules/yarn`. But it is a nightmare to extract yarn's dependencies too. We can `asarUnpack` entire `node_modules` directory but it quite the same than disabling Asar.

**Comment by chabou** [comment]
This PR now works in dev and packaged mode.
The only downside is to `asarUnpack` `app/node_modules` directory

We need to test it on Windows and Linux

**Comment by chabou** [comment]
`Yarn` is now embedded as a standalone bin, directly taken from its [release page](https://github.com/yarnpkg/yarn/releases/tag/v0.24.6). It works like a charm even with last version v0.24.6. Only this file is `asarUnpacked` 🎉 
Many thanks to @albinekb for his help.

Still need to be tested on Windows and Linux.

**Comment by ppot** [comment]
@chabou I will let you finish this because it will be cleaner for me in my third part.

**Comment by chabou** [comment]
It is finished! It only needs some testing on Windows/Linux! But I'm struggling with Win10 installation on BootCamp 😞

**Comment by albinekb** [comment]
the app/bin/yarn-standalone.js should be built in the `build` step, it shouldn't be checked into git

**Comment by chabou** [comment]
Why? I consider `yarn` as an external program. If it was written in C++, we wouldn't want to compile it by ourself. What are the pros to not have this external dependency "compiled" and versioned in our git?

**Comment by chabou** [comment]
Problem with production build on Windows solved (bad `yarn.lock` file).
Tested on Mac and Windows (dev and prod build).

Need a test on Linux

**Comment by chabou** [comment]
Tested.

Ready to be merged

File: app/plugins.js [hunk 1]
```diff
@@ -228,46 +223,47 @@ function toDependencies(plugins) {
...
-      execOptions.shell = shell;
-    }
+  function yarn(args, cb) {
+    spawnQueue.push(end => {
+      const fullcmd = [process.execPath, yarnPath].concat(args).join(' ');
+      console.log('Launching yarn:', fullcmd);
+
+      cp.exec(fullcmd, {
+        cwd: path,
+        env,
```
Comment (ppot) [inline app/plugins.js]: What happend when `isDev === true`

File: app/plugins.js [hunk 2]
```diff
@@ -228,46 +223,47 @@ function toDependencies(plugins) {
...
+        shell: true,
+        timeout: 1000 * 60 * 5,
```
Comment (ppot) [inline app/plugins.js]: add comment for `// 5min`

File: app/plugins.js [hunk 3]
```diff
@@ -228,46 +223,47 @@ function toDependencies(plugins) {
...
-
-    // https://nodejs.org/api/child_process.html#child_process_child_process_exec_command_options_callback
-    // node.js requires command line parsing should be compatible with cmd.exe on Windows, should able to accept `/d /s /c`
-    // but most custom shell doesn't. Instead, falls back to default shell
-    if (process.platform !== 'win32') {
-      execOptions.shell = shell;
-    }
+  function yarn(args, cb) {
+    spawnQueue.push(end => {
+      const fullcmd = [process.execPath, yarnPath].concat(args).join(' ');
```
Comment (ppot) [inline app/plugins.js]: `fullcmd` ?  Simply use `cmd` or `command`

File: app/plugins.js [hunk 4]
```diff
@@ -228,46 +223,47 @@ function toDependencies(plugins) {
 }
 
 function install(fn) {
-  const {shell: cfgShell, npmRegistry} = exports.getDecoratedConfig();
+  const yarnPath = resolve(__dirname, '..', 'bin', 'yarn-standalone.js');
```
Comment (ppot) [inline app/plugins.js]: `yarnPath` is not used elsewhere then inside `yarn` function. Better to move it inside function scope.

File: app/plugins.js [hunk 5]
```diff
@@ -228,46 +223,47 @@ function toDependencies(plugins) {
...
+      console.log('Launching yarn:', fullcmd);
+
+      cp.exec(fullcmd, {
+        cwd: path,
+        env,
+        shell: true,
+        timeout: 1000 * 60 * 5,
+        stdio: ['ignore', 'ignore', 'inherit']
+      }, err => {
+        console.log('Done!');
```
Comment (ppot) [inline app/plugins.js]: Be more specific. What is `Done!` ? Plugins installation? Plugins update?

File: app/plugins.js [hunk 6]
```diff
@@ -1,12 +1,13 @@
...
-
-const {app, dialog} = require('electron');
+const cp = require('child_process');
 const {sync: mkdirpSync} = require('mkdirp');
 const Config = require('electron-config');
 const ms = require('ms');
-const shellEnv = require('shell-env');
+const queue = require('queue');
+
+const spawnQueue = queue({concurrency: 1});
```
Comment (ppot) [inline app/plugins.js]: `spawnQueue` or `yarnQueue`

File: app/plugins.js (hunk 1 above)
Comment (chabou) [inline app/plugins.js]: This production flag is for `yarn`. See https://yarnpkg.com/lang/en/docs/cli/install/#toc-yarn-install-production
No side effect to Hyper.

File: app/plugins.js (hunk 6 above)
Comment (chabou) [inline app/plugins.js]: `spawnQueue` is better.

File: app/plugins.js (hunk 5 above)
Comment (chabou) [inline app/plugins.js]: We don't need this log.

File: app/plugins.js (hunk 2 above)
Comment (chabou) [inline app/plugins.js]: `ms('5m')` will be better. Thank you to have pointed this

File: app/plugins.js (hunk 3 above)
Comment (chabou) [inline app/plugins.js]: I will

File: app/plugins.js (hunk 4 above)
Comment (chabou) [inline app/plugins.js]: I will move it (`env`const too)

**Comment by chabou** [comment]
We should test plugins update too.

**Comment by chabou** [comment]
Plugins update is ok!

File: app/plugins.js (hunk 1 above)
Comment (ppot) [inline app/plugins.js]: ok!

**Comment by Stanzilla** [comment]
1. Close Hyper
2. Edit config file by hand to add a new plugin
3. Launch Hyper

For me it first errors that it can't load the new plugin, then the plugin "update" kicks in, which then installs the new plugin and then everything is fine. I just think that first error should be something we should avoid.

**Comment by chabou** [comment]
@Stanzilla this error is currently shown without my PR.
It could be better handled but it should be another PR.

**Comment by chabou** [comment]
FTR: 
`yarn` v0.24.6 is embedded and [extraResources feature](https://github.com/electron-userland/electron-builder/wiki/Options#config--platformspecificbuildoptions) of `electron-builder` is used to ship it in our bundle.

**Pull request merged**